python/ql/src/experimental/semmle/python/security/injection/LogInjection.qll

```diff
@@ -0,0 +1,24 @@
+import python
+import semmle.python.Concepts
+import experimental.semmle.python.Concepts
+import semmle.python.dataflow.new.DataFlow
+import semmle.python.dataflow.new.TaintTracking
+import semmle.python.dataflow.new.RemoteFlowSources
+
+/**
+ * A taint-tracking configuration for tracking untrusted user input used in log entries.
+ */
+class LogInjectionFlowConfig extends TaintTracking::Configuration {
+  LogInjectionFlowConfig() { this = "LogInjectionFlowConfig" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
+
+  override predicate isSink(DataFlow::Node sink) { sink = any(LogOutput logoutput).getAnInput() }
+
+  override predicate isSanitizer(DataFlow::Node node) {
+    exists(CallNode call |
+      node.asCfgNode() = call.getFunction().(AttrNode).getObject("replace") and
+      call.getArg(0).getNode().(StrConst).getText() in ["\r\n", "\n"]
+    )
+  }
```

Comment:
This is slightly unsafe as it deems either replacement ("\r\n" or "\n") a sanitizer when really both are required.
I am not sure how to improve that easily, though. Perhaps we will turn it into a concept with a single implementation and improve it over time, but I am not sure you need to do that now.

Further, this seems overly specific, so I would expect some false positives from this. Consider adding at least `StringConstCompare` as a sanitizer option. (Either directly here in this predicate, or by a construction similar to [here](https://github.com/github/codeql/blob/main/python/ql/lib/semmle/python/security/dataflow/CodeInjectionCustomizations.qll#L54).)

Reply:
Do you have a specific case here? Because the safety laboratory did not give me advice on the elimination of FP.

Reply:
If the user does `replace("\r\n", "")`, then that will match your predicate and be considered a sanitizer. However, that would not do any replacements on a message having only "\n"'s in it.

Reply:
The other comment is that if the message is compared against a known string, that should be considered a sanitizer (since that is much stronger than doing a replacement).

Reply:
I think it would be better if you give me the actual case.

Reply:
@yoff Excuse me again. Do you have examples in actual open source projects for me?

Reply:
Hm, I had to do a bit of digging. My comment came from seeing code like [this](https://lgtm.com/projects/g/mfojtik/openshift-docker-registry/snapshot/47414aaed1f5d635b9aebb1c26fba67749b325d9/files/docker_registry/images.py?sort=name&dir=ASC&mode=heatmap#L287), although in that particular case we do want to alert (because the logged value is outside the known set). I think I have seen also that only values in a known set is logged, which would be safe, and that is the sanitizer I proposed. However, it does not seem to be very common, so we could also add it once we see the first FP...